### PR TITLE
[ALLUXIO-3069] Async Caching Manager

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-3069

- Add manager to run background threads to serve "async cache" requests sent from client
- Add `getChannel` for `BlockWriter` (and its implementing classes)

### CODE REVIEW

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java [hunk 1]
```diff
@@ -11,20 +11,144 @@
...
+    synchronized (mLock) {
+      if (mPendingRequests.contains(blockId)) {
+        // This block is already planned.
+        return;
+      }
+      mPendingRequests.add(request.getBlockId());
+    }
+    mAsyncCacheExecutor.submit(() -> {
+      Protocol.OpenUfsBlockOptions openUfsBlockOptions = request.getOpenUfsBlockOptions();
+      long sessionId = IdUtils.createSessionId();
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java]: Use a negative id to indicate internal service usage?

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java [hunk 2]
```diff
@@ -11,20 +11,144 @@
...
+      long blockSize = openUfsBlockOptions.getBlockSize();
+      boolean isSourceLocal = mLocalWorkerAddress.getHost().equals(request.getSourceHost())
+          && mLocalWorkerAddress.getRpcPort() == request.getSourcePort();
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java]: `getRpcPort` -> `getDataPort`?

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java [hunk 3]
```diff
@@ -11,20 +11,144 @@
...
 
+  /**
+   * Caches the block via the local worker to read from UFS.
+   *
+   * @param sessionId session ID
+   * @param blockId block ID
+   * @param blockSize block size
+   * @param openUfsBlockOptions options to open the UFS file
+   */
+  private void cacheBlockFromUfs(long sessionId, long blockId, long blockSize,
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java]: Do we want to return boolean to indicate success/failure?

Comment (calvinjia) [reply core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java]: Also, I think we need to call something like `commitUfsBlock`, since closing the reader does not handle the block commit.

**Comment by calvinjia** [comment]
Also you can unignore the new tests added in `FileInStreamIntegrationTest` in the `async-caching` branch.

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java (hunk 2 above)
Comment (apc999) [reply core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java]: fixed

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java (hunk 1 above)
Comment (apc999) [reply core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java]: done

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java (hunk 3 above)
Comment (apc999) [reply core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java]: -  `closeUfsBlock` is added
- exceptions are thrown to indicate failures, and handled together in `submitRequest` to simplify the code. Javadoc is updated to reflect the failure cases.

**Comment by apc999** [comment]
@calvinjia  PTAL

**Comment by calvinjia** [comment]
@apc999 Thanks for the update, could you also unignore the other ignore tests ignored in async-caching? see: https://github.com/Alluxio/alluxio/pull/6606/commits/3a34713c66ad1a820434a1256c60b61f229f001a and `FileInStreamTest`

File: tests/src/test/java/alluxio/client/FileInStreamIntegrationTest.java [hunk 4]
```diff
@@ -425,48 +424,77 @@ public void positionedReadWithoutCaching() throws Exception {
...
+      FileInStream is = mFileSystem.openFile(uri, OpenFileOptions.defaults().setReadType(readType));
+      is.read();
+      URIStatus status = mFileSystem.getStatus(uri);
+      Assert.assertEquals(0, status.getInAlluxioPercentage());
+      is.close();
+      if (readType.isCache()) {
+        CommonUtils.waitFor("First block to be cached.", (input) -> {
+          try {
+            URIStatus st = mFileSystem.getStatus(uri);
+            return !st.getFileBlockInfos().get(0).getBlockInfo().getLocations().isEmpty();
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/client/FileInStreamIntegrationTest.java]: Thinking about this test again, we should also test the negative cases, ie. the other block infos have empty locations.

Comment (apc999) [reply tests/src/test/java/alluxio/client/FileInStreamIntegrationTest.java]: fixed

File: tests/src/test/java/alluxio/master/block/MultiWorkerIntegrationTest.java [hunk 5]
```diff
@@ -61,7 +60,9 @@ public void writeLargeFile() throws Exception {
             .setLocationPolicy(new RoundRobinPolicy()));
     URIStatus status = fs.getStatus(file);
     assertEquals(100, status.getInAlluxioPercentage());
-    try (FileInStream inStream = fs.openFile(file)) {
+    // Use NO_CACHE to prevent evicting "none-local" blocks
+    try (FileInStream inStream = fs
```
Comment (apc999) [inline tests/src/test/java/alluxio/master/block/MultiWorkerIntegrationTest.java]: @calvinjia , it seems if multiple workers are on the same node as the client, only one is considered as local and other workers are considered "remote" and triggering async cache. I think we should fix this (perhaps separately). what do you think?

**Comment by apc999** [comment]
@calvinjia ignored tests are fixed. PTAL

File: core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java [hunk 6]
```diff
@@ -682,63 +641,13 @@ private void testReadBuffer(int dataRead) throws Exception {
...
-      case UFS:
-        // nothing is cached, but the entire block is read
-        assertArrayEquals(new byte[0], mCacheStreams.get(index).getWrittenData());
-        assertEquals(cacheSize, mInStreams.get(index).getBytesRead());
-        break;
-      default:
-        throw new IllegalStateException("Unrecognized block source type " + mBlockSource);
-    }
+  private void validatePartialCaching(int index, int readSize) {
+    assertEquals(readSize, mInStreams.get(index).getBytesRead());
```
Comment (calvinjia) [inline core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java]: Are we still asserting that partial caching is working (ie. the block gets cached) or only that the FileInStream read behavior is correct (only reading requested bytes)? If we can validate caching that would be ideal but since this is a unit test I'm fine with changing this to validating the input stream behavior.

File: tests/src/test/java/alluxio/master/block/MultiWorkerIntegrationTest.java (hunk 5 above)
Comment (calvinjia) [reply tests/src/test/java/alluxio/master/block/MultiWorkerIntegrationTest.java]: Do you mean the behavior should be all workers are considered local, what would be involved in the fix? I don't think we have changed how we calculate locality?

File: core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java (hunk 6 above)
Comment (apc999) [reply core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java]: I prefer to go with unit-tests without verifying the async-cache for now. The reason is that
it is quite hard to mock netty RPC currently to intercept the Protos sent over netty channel.
The code will be very messy and hard to maintain. Since the async cache logic is also covered in the FileInStream integration test, I am fine with leaving this as is, until we have a better way to mock netty RPCs.

For now, a TODO is left.

File: tests/src/test/java/alluxio/master/block/MultiWorkerIntegrationTest.java (hunk 5 above)
Comment (apc999) [reply tests/src/test/java/alluxio/master/block/MultiWorkerIntegrationTest.java]: @calvinjia I think I found a better fix. Can you double check commit 201c612 and verify this is a legit fix?
Essentially, we should not async-cache if the BlockInStream is already local

**Comment by apc999** [comment]
@calvinjia , can you double check if https://github.com/Alluxio/alluxio/pull/6645/commits/18d0b687071c4642483b9dadc9cf1f13792813ee is a legit fix for `alluxio.master.block.MultiWorkerIntegrationTest#writeLargeFile`?

**Comment by calvinjia** [comment]
@apc999 Shouldn't the async cache manager automatically handle this case already, ie. if the block is read from the local worker, it will see that the block already exists and then no-op? I would prefer to not add the logic on the client side.

File: core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java (hunk 6 above)
Comment (calvinjia) [reply core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java]: Yeah I think it is better to keep the async caching logic tested in integration tests if we need to do more in the instream unit test to validate it.

**Comment by apc999** [comment]
@calvinjia , functionality wise, (a) checking the block stream source on the client or (b) leaving the check to async cache manager should be equivalent. I am curious that, since (a) saves one netty RPC from the client to its local worker, why do you prefer (b)? IMO, (a) should be strictly better than (b). Would like to hear your opinion.

**Comment by calvinjia** [comment]
@apc999 Ah I see, even though these are logically different workers, they share the same locality (hostname). Thinking about the client check again, it looks like there is no useful information lost by not doing this check on the client, so I think your change makes sense.
